src/combinatorics/simple_graph/connectivity.lean

```diff
@@ -755,4 +788,115 @@ edges_bypass_subset _
 
 end walk
 
+/-! ## `reachable` and `connected` -/
+
+/-- Two vertices are *reachable* if there is a walk between them.
+This is equivalent to `relation.refl_trans_gen` of `G.adj`.
+See `simple_graph.reachable_iff_refl_trans_gen`. -/
+def reachable (u v : V) : Prop := nonempty (G.walk u v)
+
+variables {G}
+
+lemma reachable_iff_nonempty_univ {u v : V} :
+  G.reachable u v ↔ (set.univ : set (G.walk u v)).nonempty :=
+set.nonempty_iff_univ_nonempty
+
+protected lemma reachable.elim {p : Prop} {u v : V}
+  (h : G.reachable u v) (hp : G.walk u v → p) : p :=
+nonempty.elim h hp
+
+protected lemma reachable.elim_path {p : Prop} {u v : V}
+  (h : G.reachable u v) (hp : G.path u v → p) : p :=
+begin
+  classical,
+  exact h.elim (λ q, hp q.to_path),
+end
+
+@[refl] protected lemma reachable.refl {u : V} : G.reachable u u := by { fsplit, refl }
+
+@[symm] protected lemma reachable.symm {u v : V} (huv : G.reachable u v) : G.reachable v u :=
+huv.elim (λ p, ⟨p.reverse⟩)
+
+@[trans] protected lemma reachable.trans {u v w : V}
+  (huv : G.reachable u v) (hvw : G.reachable v w) :
+  G.reachable u w :=
+huv.elim (λ puv, hvw.elim (λ pvw, ⟨puv.append pvw⟩))
+
+lemma reachable_iff_refl_trans_gen (u v : V) :
+  G.reachable u v ↔ relation.refl_trans_gen G.adj u v :=
+begin
+  split,
+  { rintro ⟨h⟩,
+    induction h,
+    { refl, },
+    { exact (relation.refl_trans_gen.single h_h).trans h_ih, }, },
+  { intro h,
+    induction h with _ _ _ ha hr,
+    { refl, },
+    { exact reachable.trans hr ⟨walk.cons ha walk.nil⟩, }, },
+end
+
+variables (G)
+
+lemma reachable_is_equivalence : equivalence G.reachable :=
+mk_equivalence _ (@reachable.refl _ G) (@reachable.symm _ G) (@reachable.trans _ G)
+
+/-- The equivalence relation on vertices given by `simple_graph.reachable`. -/
+def reachable_setoid : setoid V := setoid.mk _ G.reachable_is_equivalence
+
+/-- The connected components of a graph are elements of the quotient of vertices by
+the `simple_graph.reachable` relation. -/
+def connected_component := quot G.reachable
```

Comment:
I expect it's better to use `quotient` here since we already have the setoid?

Reply:
My understanding was that we are never meant to use `quot`? The issue with `quotient` is that setoid is a typeclass, but we have things like `quotient.mk'` to solve these things, and these have a stronger API than `quot` already; so I don't really see the advantage of `quot` here.

Reply:
I tried putting this in terms of `quotient`, but the annoying thing about it is that it puts everything in terms of `setoid.r` rather than `reachable`. Using `quot` directly avoids that.

That said, this doesn't matter when there's an specialized quotient API, and I went ahead and did that -- the only interesting part of it is `connected_component.lift_of`, which gives you an additional `is_path` assumption in case you need it.

This PR probably should be two -- I can split it if anyone wants.

Reply:
Thanks for giving this a go! Would you mind splitting it?